Add "command" package to execute external command

leshnabalara opened the pull request:
Start a async program with user set path, flags and enviornment.

leshnabalara committed:
Add "command" package to execute external command
Start a async program with user set path, flags and enviornment.

leshnabalara committed:
Remove go_binary rule in the BUILD file as suggested by gazelle

pallavagarwal07 commented on command/command.go:
```diff
@@ -0,0 +1,91 @@
+
+// ExternalCommand has the properties for invoking exec.Command
+type ExternalCommand struct {
+	path string
```
path starts with lowercase, so won't be visible to outside (users) of this package

pallavagarwal07 commented on command/command.go:
```diff
@@ -0,0 +1,91 @@
+	}
+
+	// TODO: check if client is ready
+	time.Sleep(100 * time.Millisecond)
```
This is not part of executing a program. This is independant logic, and shouldn't be combined here.

pallavagarwal07 commented on command/command.go:
```diff
@@ -0,0 +1,91 @@
+		err:       err,
+	}
+
+	stderr, err = stderrReader.ReadBytes('\n')
```
You just read one line?

pallavagarwal07 commented on command/command.go:
```diff
@@ -0,0 +1,91 @@
+}
+
+// ExecuteProgram invokes an external program
+func ExecuteProgram(command ExternalCommand, stdout []byte, stderr []byte, status chan Status) error {
```
Instead of returning error, create a struct that has methods defined on it for status, StdoutNonBlocking StderrNonBlocking (which return (all output, err) at the time they are called. you only read any available data _at_ the time this method is called and not use a goroutine polling it all the time).

In the arguments, in ExternalCommand, take a callback to be called with error once program ends.

pallavagarwal07 commented on command/command_test.go:
```diff
@@ -0,0 +1,48 @@
+	var errstream []byte
+
+	xephyr := ExternalCommand{
+		path: "Xephyr",
```
Please use an independent binary for testing features. That binary will be a data dependency (look at data deps in bazel) of this test, and you'd use runfiles to get the path to that binary.

leshnabalara committed:
Provide wrapper around exec.Cmd
add functionality to get non-blocking stderr and stdout

leshnabalara committed:
Add async close of external program
Add tests using cat command

pallavagarwal07 commented on command/command.go:
```diff
@@ -0,0 +1,85 @@
+func ExecuteProgram(command ExternalCommand, errorHandler func(*exec.Cmd, error) error) (*ProgramState, error) {
+	var err error
+	programState := &ProgramState{}
+	cmd := exec.Command(
```
nit: Indent looks weird, just put arguments in same line?

pallavagarwal07 commented on command/command.go:
```diff
@@ -0,0 +1,85 @@
+
+	programState.StderrNonBlocking()
+	programState.StdoutNonBlocking()
+	if err := cmd.Wait(); err != nil {
```
Call the callback even if err is nil (it signified end of execution of binary)

pallavagarwal07 commented on command/command.go:
```diff
@@ -0,0 +1,85 @@
+	programState.StderrNonBlocking()
+	programState.StdoutNonBlocking()
+	if err := cmd.Wait(); err != nil {
+		errorHandler(cmd, err)
```
I wonder if programState is the better object to return here?

pallavagarwal07 commented on command/command.go:
```diff
@@ -0,0 +1,85 @@
+
+func (p *ProgramState) StdoutNonBlocking() ([]byte, error) {
+
+	stdout, err := ioutil.ReadAll(p.stdoutPipe)
```
Wouldn't this block until program exits or closes the stdout pipe?

pallavagarwal07 commented on command/command.go:
```diff
@@ -0,0 +1,85 @@
+
+func (p *ProgramState) StderrNonBlocking() ([]byte, error) {
+
+	stderr, err := ioutil.ReadAll(p.stderrPipe)
```
This and above function are identical except for stdout/stderr things which could be arguments. Separate out the logic in another function.

pallavagarwal07 commented on command/command_test.go:
```diff
@@ -0,0 +1,44 @@
+func TestExecuteProgram(t *testing.T) {
+
+	cat := ExternalCommand{
+		Path: "cat",
```
Please use your own binary that:

When started, prints some stuff on stdout/err but doesn't exit.

Has a signal handler that catches signal sent by test to print some more stuff

Has a signal handler that catches a _different_ signal sent by test and exits.

leshnabalara committed:
Improve command_test.go by using programmed binary
Add test_sample_program.go as a binary to test command.go
Remove redundant NonBlockingCall code by breaking it down into another
function.

leshnabalara committed:
Add mutex for safe use of err and output slices
Ensure wait() is not executed until reading from pipe stops
provide getters for stderr and stdout

pallavagarwal07 commented on command/command.go:
```diff
@@ -0,0 +1,126 @@
+	go func() {
+		errStdout := programState.StdoutListner()
+		if errStdout != nil {
+			errorHandler(programState, errStdout)
```
errorHandler is only for the case of program exiting. Documentation says that this error can _only_ be eof, so just ignore it, and stop reading from this pipe.

pallavagarwal07 commented on command/command.go:
```diff
@@ -0,0 +1,126 @@
+
+	errStderr := programState.StderrListner()
+	if errStderr != nil {
+		errorHandler(programState, errStderr)
```
Same as above

pallavagarwal07 commented on command/command.go:
```diff
@@ -0,0 +1,126 @@
+
+func (p *ProgramState) Stdout() []byte {
+	p.stdoutmutex.Lock()
+	stdout := p.stdout
```
This actually just stores a reference to the same slice. you'll have to duplicate it instead (there would be some helper function for it).

pallavagarwal07 commented on command/command.go:
```diff
@@ -0,0 +1,126 @@
+
+func (p *ProgramState) Stderr() []byte {
+	p.stderrmutex.Lock()
+	stderr := p.stderr
```
Same here

pallavagarwal07 commented on command/command_test.go:
```diff
@@ -0,0 +1,52 @@
+func TestExecuteProgram(t *testing.T) {
+
+	test := ExternalCommand{
+		Path: "./test_command",
```
Are you trying to execute a directory? The path to the binary is test_command/test_sample_program, right?

pallavagarwal07 commented on command/command_test.go:
```diff
@@ -0,0 +1,52 @@
+	}
+
+	want := "Hello"
+	got := string(out)
```
See, it takes some time for your test program to actually output anything, so even here, you need to wait a while to check if the output is on the way.

Usually, the right way is to check the value of out() in a loop, till some timeout (you can have a helper that does it).

pallavagarwal07 replied:
Something like this:

```
timeout := time.After(1 * time.Second)

for want != got {
    select {
    case <-timeout:
        t.Errorf("Got text = %q, want %q", got, want)
    default:
    }
    got = string(programstate.Stdout())
}
```

pallavagarwal07 commented on command/command.go:
```diff
@@ -0,0 +1,126 @@
+
+		buf, err := reader.ReadBytes('\n')
+		mutex.Lock()
+		buffer = append(buffer, buf...)
```
This is a bit tricky. So even though slices are references to an underlying array, it is not guaranteed that append with change the same slice. It can also return a new slice. Which means you need to take in a pointer to a slice, and assign to *buffer.

pallavagarwal07 commented on command/command_test.go:
```diff
@@ -0,0 +1,52 @@
+		fmt.Println("Returned error by StderrNonBlocking()", string(serr))
+	}
+
+	want := "Hello"
```
You actually want "Hello\n"

leshnabalara committed:
Ignore eof errors in err/out listner
return a copy of stderr and stdout in getters

leshnabalara committed:
Generate absolute path for test command using bazel.Runfile()

pallavagarwal07 commented on command/command_test.go:
```diff
@@ -0,0 +1,77 @@
+	want := "Hello\n"
+	got := testWaitOutput(programstate.Stdout, want)
+
+	serr := programstate.Stderr()
```
Combine this with line below.

Also, you shouldn't have println or logging statement in tests. If it is something that shouldn't happen, use t.error

pallavagarwal07 commented on command/command_test.go:
```diff
@@ -0,0 +1,77 @@
+		t.Errorf("Got text = %q, want %q", got, want)
+	}
+
+	err = programstate.Command.Process.Signal(os.Interrupt)
```
Combine with if statement

pallavagarwal07 commented on command/command_test.go:
```diff
@@ -0,0 +1,77 @@
+	want = "Hello\nWorld\n"
+	got = testWaitOutput(programstate.Stdout, want)
+
+	serr = programstate.Stderr()
```
Combine with if

pallavagarwal07 commented on command/command_test.go:
```diff
@@ -0,0 +1,77 @@
+	}
+
+	want = "Hello\nWorld\n"
+	got = testWaitOutput(programstate.Stdout, want)
```
This works, but a cleaner way would be to modify the function to return err == nil if the output does match within timeout and a new error with the error message if it doesn't. then move this function & combine it with if statement.
Right now both your loop, and the got != want below are checking for content equality.

leshnabalara committed:
Improve formatting
